src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,411 @@
+    def messages(self):
+        return self.__messages
+
+    def append(self, message: Message[KafkaPayload]) -> None:
+        self.__messages.append(message)
+
+    def ready(self) -> bool:
+        if len(self.messages) >= self.__max_batch_size:
+            return True
+        elif time.time() >= self.__deadline:
+            return True
+        else:
+            return False
+
+
+class BatchMessages(ProcessingStep[KafkaPayload]):
+    """
+    First processing step in the MetricsConsumerStrategyFactory.
+    Keeps track of a batch of messages (using the MetricsBatchBuilder)
+    and then when at capacity, either max_batch_time or max_batch_size,
+    flushes the batch.
+
+    Flushing the batch here means wrapping the batch in a Message, the batch
+    itself being the payload. This is what the ParallelTransformStep will
+    process in the process_message function.
+    """
+
+    def __init__(
+        self,
+        next_step: ProcessingStrategy[MessageBatch],
+        max_batch_time: float,
+        max_batch_size: int,
+    ):
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+
+        self.__next_step = next_step
+        self.__batch: Optional[MetricsBatchBuilder] = None
+        self.__closed = False
+
+    def poll(self) -> None:
+        assert not self.__closed
+
+        self.__next_step.poll()
+
+        if self.__batch and self.__batch.ready():
+            self.__flush()
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        if self.__batch is None:
+            self.__batch = MetricsBatchBuilder(self.__max_batch_size, self.__max_batch_time)
+
+        self.__batch.append(message)
+
+    def __flush(self) -> None:
+        if not self.__batch:
+            return
+        last = self.__batch.messages[-1]
+
+        new_message = Message(last.partition, last.offset, self.__batch.messages, last.timestamp)
+
+        self.__next_step.submit(new_message)
+        self.__batch = None
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        self.__flush()
+        self.__next_step.join(timeout)
+
+
+class ProducerResultFuture(NamedTuple):
+    message: Message[KafkaPayload]
+    future: Future
+
+
+class ProduceStep(ProcessingStep[MessageBatch]):
+    """
+    Step that produces to the snuba-metrics topic, collecting the futures returned by
+    the producer. Continously checks to see if futures are done, and once that's the case
+    can commit up to the last future that is done.
+    """
+
+    def __init__(self, commit_function: Callable[[Mapping[Partition, Position]], None]) -> None:
+        snuba_metrics = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS]
+        snuba_metrics_producer = KafkaProducer(
+            kafka_config.get_kafka_producer_cluster_options(snuba_metrics["cluster"]),
+        )
```

Comment:
@fpacifici I added the option to pass in the producer as optional so I could use that for a testing producer, but I am still creating the producer in the init if it's not passed in. It makes sense to me to pass in the producer regardless. In that case would I have to do something like the following? https://github.com/getsentry/sentry/blob/a1aa63d61ee63340bc550a2b1da5f5b701c886f7/src/sentry/runner/commands/run.py#L402-L408

Reply:
It makes sense to pass the consumer instance instead of recreating it. 
Though feel free to do that in a follow up PR, it is not critical.
In that case you will have to close the producer right before exiting.
It is like what the executor does:
https://github.com/getsentry/snuba/blob/master/snuba/cli/subscriptions_executor.py#L110-L128
Line 127 specifically